notebooks/official/ray_on_vertex_ai/spark_on_ray_on_vertex_ai.ipynb

```diff
@@ -0,0 +1,1470 @@
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "a9bd291fca90"
+      },
+      "outputs": [],
+      "source": [
+        "%%writefile {SCRIPT_DIR}/pandas_udf_local_dependency.py\n",
+        "\n",
+        "import pandas as pd\n",
+        "import pyspark\n",
+        "import raydp\n",
+        "from pyspark.sql.functions import pandas_udf\n",
+        "from pyspark.sql.types import StringType\n",
+        "\n",
+        "def test_udf(spark: pyspark.sql.SparkSession):\n",
+        "    import pandas as pd\n",
+        "    df = spark.createDataFrame(pd.read_csv(\"https://www.datavis.ca/gallery/guerry/guerry.csv\"))\n",
+        "    import pathlib\n",
+        "    module_path = str(pathlib.Path(__file__).parent.resolve())\n",
+        "    return df.select(udf_wrapper_func('Lottery','Literacy', 'Pop1831', module_path)).collect()\n",
+        "\n",
+        "def udf_wrapper_func(s1: pd.Series, s2: pd.Series, s3: pd.Series, module_path: str) -> str:\n",
+        "\n",
+        "    @pandas_udf(StringType())\n",
+        "    def func(s1: pd.Series, s2: pd.Series, s3: pd.Series) -> str:\n",
+        "        import sys\n",
+        "        sys.path.append(module_path)\n",
+        "        \n",
+        "        # import local module\n",
+        "        import my_module\n",
+        "        my_module.print_func(\"This is a UDF local dependency test.\")\n",
+        "\n",
+        "        import numpy as np\n",
+        "        import subprocess\n",
+        "        import sys\n",
+        "        subprocess.check_call([sys.executable, \"-m\", \"pip\", \"install\", \"statsmodels\"])\n",
+        "        import statsmodels.api as sm\n",
+        "        import statsmodels.formula.api as smf\n",
+        "\n",
+        "        d = {'Lottery': s1, \n",
+        "             'Literacy': s2,\n",
+        "             'Pop1831': s3}\n",
+        "        data = pd.DataFrame(d)\n",
+        "\n",
+        "    #     # Fit regression model (using the natural log of one of the regressors)\n",
+        "        results = smf.ols('Lottery ~ Literacy + np.log(Pop1831)', data=data).fit()\n",
+        "        return results.summary().as_csv()\n",
+        "\n",
+        "    return func(s1, s2, s3)\n",
+        "\n",
+        "\n",
+        "if __name__ == '__main__':\n",
+        "    \n",
+        "    spark = raydp.init_spark(\n",
+        "      app_name=\"UDF_TEST\",\n",
+        "      num_executors=2,\n",
+        "      executor_cores=2,\n",
+        "      executor_memory=\"500M\",\n",
+        "    )\n",
+        "    \n",
+        "    print(test_udf(spark))\n",
+        "    \n",
+        "    raydp.stop_spark()"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "90cb909f991b"
+      },
+      "source": [
+        "#### Submit the job"
```

Comment:
Is it worth creating a helper function for this code to reuse in the steps above?